### ISSUE TITLE
@typescript-eslint/parser v4: incorrect lints from plugin eslint-plugin-react-hooks with generics

### ISSUE THREAD

**Issue by NeoRaider** [typescript-eslint/typescript-eslint#2467]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

I'm not quite sure where to report this issue, as the problem lies in the interaction between two independently developed packages. As I started seeing this when updating @typescript-eslint/parser from 3.7.0 to 4.0.1, I'll try here first...

**Repro**

```JS
module.exports = {
        parser: '@typescript-eslint/parser',
        plugins: [
                'react-hooks',
        ],
        parserOptions: {
                ecmaVersion: 2018,
                sourceType: 'module',
        },
        rules: {
                'react-hooks/exhaustive-deps': 'warn',
        },
        settings: {
                react: {
                        version: 'detect',
                },
        },
};
```

```TS
export function useFoo<T>(): (foo: T) => boolean {
        return useCallback((foo: T) => false, []);
}

```

**Expected Result**

Nothing

**Actual Result**

The `react-hooks/exhaustive-deps` is emitting a warning about a missing dependency, giving me the impossible suggestion to add the generic parameter `T` to the dep array.

```
  2:40  warning  React Hook useCallback has a missing dependency: 'T'. Either include it or remove the dependency array  react-hooks/exhaustive-deps
```

**Additional Info**

Reproducer repo at: https://github.com/NeoRaider/eslint-bug-reproducer

```
yarn install
yarn lint
```

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `4.0.1` |
| `TypeScript`                | `4.0.2` |
| `ESLint`                    | `7.8.0` |
| `node`                      | `14.9.0` |
| `eslint-plugin-react-hooks` | `4.0.8` |

**Comment by bradzacher** [comment]
You should raise a bug with `eslint-plugin-react-hooks` for this.

The rule needs to be updated to be aware of type variables. Right now it assumes that all variables are value variables, and all references are value references, but that is not the case.
